File: backend/src/main/scala/domains/usergroup/objects/UserGroupRole.scala
package domains.usergroup.objects

import io.circe.{Decoder, Encoder}


/** 用户组成员角色，owner 代表所有者，manager 代表管理成员，member 为普通成员。 */
enum UserGroupRole:
  case Owner
  case Manager
  case Member

/** 提供用户组成员角色的 JSON 编解码和输入解析。 */
object UserGroupRole:
  given Encoder[UserGroupRole] = Encoder.encodeString.contramap(encode)
  given Decoder[UserGroupRole] = Decoder.decodeString.emap(parse)

  /** 解析组内角色字符串，非法值返回业务校验错误。 */
  def parse(value: String): Either[String, UserGroupRole] =
    value.trim match
      case "owner" => Right(UserGroupRole.Owner)
      case "manager" => Right(UserGroupRole.Manager)
      case "member" => Right(UserGroupRole.Member)
      case _ => Left("User group role must be one of: owner, manager, member.")

  private def encode(value: UserGroupRole): String =
    value match
      case UserGroupRole.Owner => "owner"
      case UserGroupRole.Manager => "manager"
      case UserGroupRole.Member => "member"

File: frontend/src/objects/usergroup/UserGroupRole.ts
/** 用户组成员角色；owner 为所有者，manager 可管理成员，member 为普通成员。 */
export type UserGroupRole = 'owner' | 'manager' | 'member'

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

/** 解析用户组角色；拒绝未知角色字符串。 */
export function parseUserGroupRole(rawRole: string): ParseResult<UserGroupRole> {
  if (rawRole === 'owner' || rawRole === 'manager' || rawRole === 'member') {
    return { ok: true, value: rawRole }
  }

  return { ok: false, error: 'Unknown user group role.' }
}
